iroha_p2p/src/network.rs

```diff
@@ -0,0 +1,154 @@
+use std::{collections::HashMap, io::Error};
+
+use async_stream::stream;
+use futures::Stream;
+use iroha_actor::{broker::Broker, Actor, Addr, Context, ContextHandler, Handler};
+use iroha_crypto::PublicKey;
+use iroha_data_model::peer::prelude::PeerId;
+use iroha_logger::warn;
+use tokio::net::{TcpListener, TcpStream};
+use ursa::{encryption::symm::Encryptor, kex::KeyExchangeScheme};
+
+use crate::peer::{Peer, State};
+
+/// Main network layer structure, that is holding connections, called [`Peer`]s.
+#[derive(Debug)]
+pub struct Network<K: KeyExchangeScheme + Send + 'static, E: Encryptor + Send + 'static> {
+    peers: HashMap<PeerId, Addr<Peer<K, E>>>,
+    listener: Option<TcpListener>,
+    broker: Broker,
+    //handler: AlwaysAddr<A>
+}
+
+impl<K: KeyExchangeScheme + Send + 'static, E: Encryptor + Send + 'static> Network<K, E> {
+    /// Creates a network structure, that will hold connections to other nodes.
+    pub fn new(broker: Broker /*handler: AlwaysAddr<A>*/) -> Self {
+        Self {
+            peers: HashMap::new(),
+            listener: None,
+            broker,
+            /*handler*/
+        }
+    }
+
+    fn listener_stream(&mut self) -> impl Stream<Item = NewPeerResult> + Send + 'static {
+        #[allow(clippy::expect_used)]
+        let listener = self.listener.take().expect("Unexpected");
+        stream! {
+            loop {
+                match listener.accept().await {
+                    Ok((stream, addr)) => {
+                        let id = PeerId { address: addr.to_string(), public_key: PublicKey::default() };
```

Comment:
Public key shouldn't be default here

Reply:
We don't know it's public key right on connection. The peer will somehow send it to us after handshakes.